File: mcp/src/clients/graphql.ts
const GRAPHQL_URL = "https://index.bostrom.cybernode.ai/v1/graphql";

export async function graphql<T = unknown>(
  query: string,
  variables?: Record<string, unknown>,
): Promise<T> {
  const res = await fetch(GRAPHQL_URL, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ query, variables }),
  });
  if (!res.ok) {
    throw new Error(`GraphQL ${res.status}: ${res.statusText}`);
  }
  const json = (await res.json()) as { data?: T; errors?: Array<{ message: string }> };
  if (json.errors?.length) {
    throw new Error(`GraphQL error: ${json.errors[0].message}`);
  }
  return json.data as T;
}

File: mcp/rust/src/clients/graphql.rs
use anyhow::{Context, Result};
use serde_json::Value;
use std::time::Duration;

#[derive(Clone)]
pub struct GraphqlClient {
    client: reqwest::Client,
    url: String,
}

impl GraphqlClient {
    pub fn new(url: &str) -> Self {
        let client = reqwest::Client::builder()
            .timeout(Duration::from_secs(30))
            .build()
            .expect("failed to build HTTP client");
        Self {
            client,
            url: url.to_string(),
        }
    }

    /// Execute a GraphQL query, returning `.data` from the response.
    pub async fn query(&self, query: &str, variables: Option<&Value>) -> Result<Value> {
        let body = serde_json::json!({
            "query": query,
            "variables": variables.unwrap_or(&Value::Null),
        });
        let resp = self
            .client
            .post(&self.url)
            .json(&body)
            .send()
            .await
            .with_context(|| "GraphQL POST")?;
        let status = resp.status();
        if !status.is_success() {
            let body = resp.text().await.unwrap_or_default();
            anyhow::bail!("GraphQL returned {status}: {body}");
        }
        let json: Value = resp.json().await.context("GraphQL json parse")?;
        if let Some(errors) = json.get("errors") {
            if errors.is_array() && !errors.as_array().unwrap().is_empty() {
                anyhow::bail!("GraphQL errors: {errors}");
            }
        }
        Ok(json.get("data").cloned().unwrap_or(json))
    }
}
